bugsnag-android-core/src/main/java/com/bugsnag/android/Event.kt

```diff
@@ -0,0 +1,174 @@
+     * @return the grouping hash, if set
+     */
+    /**
+     * Set a custom grouping hash to use when grouping this Event on the
+     * Bugsnag dashboard. By default, we use a combination of error class
+     * and top-most stacktrace line to calculate this, and we do not recommend
+     * you override this.
+     */
+    var groupingHash: String? = null
+
+    /**
+     * Get the context associated with this Event.
+     */
+    /**
+     * Override the context sent to Bugsnag with this Event. By default we'll
+     * attempt to detect the name of the top-most Activity when this error
+     * occurred, and use this as the context, but sometimes this is not
+     * possible.
+     */
+    var context: String? = null
+
+    /**
+     * @return user information associated with this Event
+     */
+    internal var _user = User(null, null, null)
+
+    protected fun shouldIgnoreClass(): Boolean {
+        return config.ignoreClasses.contains(errors[0].errorClass)
+    }
+
+    @Throws(IOException::class)
+    override fun toStream(writer: JsonStream) {
+        // Write error basics
+        writer.beginObject()
+        writer.name("context").value(context)
+        writer.name("metaData").value(metadata)
+
+        writer.name("severity").value(severity)
+        writer.name("severityReason").value(handledState)
+        writer.name("unhandled").value(handledState.isUnhandled)
+
+        writer.name("projectPackages").beginArray()
+        projectPackages.forEach { writer.value(it) }
+        writer.endArray()
+
+        // Write exception info
+        writer.name("exceptions")
+        writer.beginArray()
+        errors.forEach { writer.value(it) }
+        writer.endArray()
+
+        // Write user info
+        writer.name("user").value(_user)
+
+        // Write diagnostics
+        writer.name("app").value(app)
+        writer.name("device").value(device)
+        writer.name("breadcrumbs").value(breadcrumbs)
+        writer.name("groupingHash").value(groupingHash)
+
+        writer.name("threads")
+        writer.beginArray()
+        for (thread in threads) {
+            writer.value(thread)
+        }
+        writer.endArray()
+
+        if (session != null) {
+            val copy = Session.copySession(session)
+            writer.name("session").beginObject()
+            writer.name("id").value(copy.id)
+            writer.name("startedAt").value(DateUtils.toIso8601(copy.startedAt))
+
+            writer.name("events").beginObject()
+            writer.name("handled").value(copy.handledCount.toLong())
+            writer.name("unhandled").value(copy.unhandledCount.toLong())
+            writer.endObject()
+            writer.endObject()
+        }
+
+        writer.endObject()
+    }
+
+    protected fun updateSeverityInternal(severity: Severity) {
+        handledState = HandledState.newInstance(handledState.severityReasonType,
+            severity, handledState.attributeValue)
+        this.severity = severity
+    }
```

Comment:
If the user alters the severity on an `Event`, the `severityReason` is updated to `userCallbackSetSeverity` and a different explanation is shown on the dashboard.

We need a separate method to update the severity after an `Event` is created to prevent the `severityReason` changing when [notifying of a Native error](https://github.com/bugsnag/bugsnag-android/blob/event-interface/bugsnag-android-core/src/main/java/com/bugsnag/android/NativeInterface.java#L390).

Reply:
It would be useful to add some Javadoc to explain this.